Jasmine/Tetris/ViewModels/TetrisGameViewModel.swift

```diff
@@ -1,4 +1,200 @@
+    private func getTileText(at coordinate: Coordinate) -> String {
+        return tetrisGrid.getTileText(at: coordinate) ?? " "
+    }
+
+    /// Shifts all the tiles above `coordinates` 1 row down.
+    /// Starts from the row right above the coordinates so that it can break once an empty tile is encountered
+    fileprivate func shiftDownTiles(_ coordinates: Set<Coordinate>) {
+        var coordinatesToShift: [(from: Coordinate, to: Coordinate)] = []
+        for coordinate in coordinates {
+            for row in (0..<coordinate.row).reversed() {
+                let currentCoordinate = Coordinate(row: row, col: coordinate.col)
+                guard let text = tetrisGrid.removeTile(at: currentCoordinate) else {
+                    break
+                }
+                let newCoordinate = currentCoordinate.nextRow
+                tetrisGrid.addTile(at: newCoordinate, tileText: text)
+                coordinatesToShift.append((from: currentCoordinate, to: newCoordinate))
+            }
+        }
+        delegate?.animate(shiftTiles: coordinatesToShift)
+    }
+
+    // TODO: fetch from database to match valid phrase
+    private func getValidPhraseRange(_ line: String) -> CountableRange<Int>? {
+        let phrases = ["先发制人"]
+        for phrase in phrases {
+            let phraseLen = phrase.characters.count
+            for i in 0...line.characters.count - phraseLen {
+                let startIndex = line.index(line.startIndex, offsetBy: i)
+                if line[startIndex..<line.index(startIndex, offsetBy: phraseLen)] == phrase {
+                    return i..<i + phraseLen
+                }
+            }
+        }
+        return nil
+    }
+
+    // TODO: generate from database, base on existing grid
+    fileprivate func getRandomWord() -> String {
+        let words = "先发制人"
+        return String(words[words.index(words.startIndex,
+                                        offsetBy: Random.integer(toInclusive: words.characters.count))])
+    }
+}
+
+extension TetrisGameViewModel: TetrisGameViewModelProtocol {
+
+    var gameTitle: String {
+        return Constants.Game.Tetris.gameTitle
+    }
+
+    var gameInstruction: String {
+    	return Constants.Game.Tetris.gameInstruction
+    }
+
+    func canShiftFallingTile(to coordinate: Coordinate) -> Bool {
+        return !tetrisGrid.hasTile(at: coordinate)
+    }
+
+    func dropNextTile() -> (location: Coordinate, tileText: String) {
+        upcomingTiles.append(getRandomWord())
+        let tileText = upcomingTiles.removeFirst()
+        delegate?.redisplayUpcomingTiles()
+
+        fallingTileText = tileText
+        let randCol = Random.integer(toInclusive: Constants.Game.Tetris.columns)
+        return (location: Coordinate(row: Coordinate.origin.row, col: randCol),
+                tileText: tileText)
+    }
+
+    func landFallingTile(at coordinate: Coordinate) {
+        guard let fallingTileText = fallingTileText else {
```

Comment:
Can't change to `String` since it's only initialized after VC command it to `dropNextTile`. 

> Implicitly unwrapped optionals are useful when an optional’s value is confirmed to exist immediately after the optional is first defined and can definitely be assumed to exist at every point thereafter. The primary use of implicitly unwrapped optionals in Swift is during class initialization, as described in Unowned References and Implicitly Unwrapped Optional Properties.

In this case, the variable is only initialized much later when VC calls `dropNextTile`, so i don't think it's good styling to implicitly unwrap it.